### PR TITLE
🔧 Adding support for graceful shutdowns #2 

### PR DESCRIPTION
**Pull request by josemkorivo**

### CODE REVIEW

**Comment by dumindu** [comment]
@josemkorivo Thanks for the PR 👏
This weekend I am little bit busy and a long weekend is coming next week. So let me check your PR next week due to I am using a little bit different graceful shutdown implementation on some of private projects.
Thanks and enjoy your weekend!

**Comment by josemkorivo** [comment]
Thanks, Dumindu for the update, will be grateful if I get your feedback. I
am a Junior, just got started with go, and it's my first pull request as
well.

File: cmd/app/main.go [hunk 1]
```diff
@@ -44,7 +48,43 @@ func main() {
 		IdleTimeout:  appConf.Server.TimeoutIdle,
 	}
 
-	if err := s.ListenAndServe(); err != nil && err != http.ErrServerClosed {
-		logger.Fatal().Err(err).Msg("Server startup failed")
+	// Make a channel to listen for errors coming from the listener. Use a buffered
```
Comment (dumindu) [inline cmd/app/main.go:51]: I would like to not add much comments on code, unless it's really necessary.

File: cmd/app/main.go [hunk 2]
```diff
@@ -44,7 +48,43 @@ func main() {
...
+	// Block main until signal is received
+	select {
+	case err := <-serverErrors:
+		logger.Info().Msgf("error : listening and serving %s", err)
+
+	case <-shutdown:
+		logger.Info().Msg("main : Start shutdown")
+
+		// Give outstanding requests a deadline for completion
+		const timeout = 30 * time.Second
```
Comment (dumindu) [inline cmd/app/main.go:74]: 30s should be in appConf.Server. May be you can use appConf.Server.TimeoutIdle directly.

**Comment by dumindu** [comment]
@josemkorivo Hey buddy,
I am so sorry, I was quite busy with some another project. 
However I could check the PR a bit and need to do few small modifications like using `syscall.SIGTERM` instead `os.Kill`, stopping deps like db and few more typo fixes to make the code more readability. I will do the modifications and will double check and will push this PR within upcoming week. 
So sorry again and thanks for the PR.

File: cmd/app/main.go [hunk 3]
```diff
@@ -44,7 +47,35 @@ func main() {
 		IdleTimeout:  appConf.Server.TimeoutIdle,
 	}
 
-	if err := s.ListenAndServe(); err != nil && err != http.ErrServerClosed {
-		logger.Fatal().Err(err).Msg("Server startup failed")
+	serverErrors := make(chan error, 1)
+
+	go func() {
```
Comment (dumindu) [inline cmd/app/main.go:52]: @josemkorivo Hi, 
Btw why you need to run `ListenAndServe()` in a separate goroutine? From where you see this pattern? 
As I know, it's a blocking and has bundled with the port we given. I think there is no special benefit on this.

**Comment by dumindu** [comment]
@josemkorivo Sorry for the delay.
- Due to we already separated executables under `/cmd`, we don't need to run `ListenAndServe()` in a separate Goroutine. New changes were based on https://golang.org/pkg/net/http/#example_Server_Shutdown . However I tried to double check your suggestions even in https://www.reddit.com/r/golang/comments/i6fopj and came up with the current solution to make the code more simpler.
- Instead using `os.Kill`, `syscall.SIGTERM` were added.

Sorry for the delay again and thanks for the contribution 👏

**Comment by josemkorivo** [comment]
no problem, it's also easier to understand especially for people who are
just getting started.

**Pull request merged**